sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -17,36 +17,69 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl;
 
+import java.util.Map;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Internal;
 import org.apache.beam.sdk.extensions.sql.BeamSql;
 import org.apache.beam.sdk.extensions.sql.BeamSqlCli;
+import org.apache.beam.sdk.extensions.sql.BeamSqlTable;
 import org.apache.beam.sdk.extensions.sql.BeamSqlUdf;
 import org.apache.beam.sdk.extensions.sql.impl.interpreter.operator.UdafImpl;
-import org.apache.beam.sdk.extensions.sql.impl.planner.BeamQueryPlanner;
+import org.apache.beam.sdk.extensions.sql.meta.provider.ReadOnlyTableProvider;
 import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
+import org.apache.beam.sdk.extensions.sql.meta.store.InMemoryMetaStore;
 import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.Row;
 import org.apache.calcite.jdbc.CalciteConnection;
 import org.apache.calcite.jdbc.CalcitePrepare;
+import org.apache.calcite.plan.RelOptUtil;
 import org.apache.calcite.schema.SchemaPlus;
 import org.apache.calcite.schema.impl.ScalarFunctionImpl;
+import org.apache.calcite.sql.SqlExecutableStatement;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.tools.RelConversionException;
+import org.apache.calcite.tools.ValidationException;
 
 /**
  * {@link BeamSqlEnv} prepares the execution context for {@link BeamSql} and {@link BeamSqlCli}.
  *
  * <p>It contains a {@link SchemaPlus} which holds the metadata of tables/UDF functions, and a
  * {@link BeamQueryPlanner} which parse/validate/optimize/translate input SQL queries.
  */
+@Internal
+@Experimental
 public class BeamSqlEnv {
   final CalciteConnection connection;
   final SchemaPlus defaultSchema;
   final BeamQueryPlanner planner;
 
-  public BeamSqlEnv(TableProvider tableProvider) {
+  private BeamSqlEnv(TableProvider tableProvider) {
```

Comment:
I disagree with this part of the refactor. I don't think the details of the TableProvider should be inside BeamSqlEnv. The details of how the table provider is created and what it contains are very user specific.

Reply:
In general case the details of `TableProvider` are not inside of `BeamSqlEnv`, there's `withTableProvider()` to use with any custom table provider.

The intention here was to have a few convenience methods for most common scenarios to reduce confusion a bit. Otherwise it's hard to understand what `MetaStore` or table provider I should use and how without copy-pasting from some other example code.

Another approach I am thinking is to just leave the `withTableProvider()` and probably add factory methods to the table providers instead. What do you think?

Reply:
Adding factory methods to the table providers sounds like a great option! All `withTableProvider()` does is call the constructor with the exact same set of arguments. What value does keeping it provide?

Reply:
In this case it doesn't add too much value, except for having an explicit name which tells the reader what it does. My understanding is that in Beam we prefer factory methods over constructors, @kennknowles mentioned to consider the factory method for `BeamSqlEnv` in another PR.

In general, the argument of factory methods vs constructors in Java goes like this (from [Effective Java](http://www.informit.com/articles/article.aspx?p=1216151)):

Pros of factory methods:
 - they have names;
 - they're not required to create new instances;
 - can return subtype;
 - better type inference for generics;

Cons:
 - prohibit inheritance if there are no not-private constructors;
 - hard to find in docs;

Reply:
Pro:

 - prohibit inheritance :-) and encourage composition instead

Reply:
I hear adding `final` to the class prohibits inheritance without needing a redundant function. I definitely see the benefits of factory methods over multiple constructors. This is a really minor thing, so ignore my comment and do what you think is right here.

Reply:
Kept them for now since no strong opinions about it